capsules/src/ble.rs

```diff
@@ -0,0 +1,287 @@
+}
+// 'a = lifetime
+// R - type Radio
+impl<'a, R: BleAdvertisementDriver + 'a, A: hil::time::Alarm + 'a> BLE<'a, R, A> {
+    pub fn new(radio: &'a R,
+               container: Container<App>,
+               buf: &'static mut [u8],
+               alarm: &'a A)
+               -> BLE<'a, R, A> {
+        BLE {
+            radio: radio,
+            busy: Cell::new(false),
+            app: container,
+            kernel_tx: TakeCell::new(buf),
+            alarm: alarm,
+            // 5017 : every 150 ms!?
+            // how is that comptued?
+            // 1 clock cycle, 1/(16*10^6) = 6.25e-8
+            // 5007 * 6.25e-8 ~= 0.31ms
+            // TODO: check this if other CPUs shall be supported
+            interval: Cell::new(5017),
+            advertise: Cell::new(false),
+            // 6 bytes for 'TockOS'
+            // third byte is zero
+            offset: Cell::new(0),
+        }
+    }
+
+    pub fn set_adv_data(&self, ad_type: usize) -> ReturnCode {
+        for cntr in self.app.iter() {
+            cntr.enter(|app, _| {
+                app.app_write
+                    .as_ref()
+                    .map(|slice| {
+                        let len = slice.len();
+                        let i = self.offset.get() + len + 2;
+                        if i < 31 {
+                            self.kernel_tx
+                                .take()
+                                .map(|data| {
+                                    for (out, inp) in data.iter_mut()
+                                        .zip(slice.as_ref()[0..len].iter()) {
+                                        *out = *inp;
+                                    }
+                                    let tmp = self.radio
+                                        .set_adv_data(ad_type, data, len, self.offset.get() + 9);
+                                    self.kernel_tx.replace(tmp);
+                                    self.offset.set(i);
+                                });
+                        }
+                    });
+            });
+        }
+        ReturnCode::SUCCESS
+    }
+
+    pub fn configure_periodic_alarm(&self) {
+        self.radio.set_channel(37);
+        let tics = self.alarm.now().wrapping_add(5017 as u32);
+        self.alarm.set_alarm(tics);
+    }
+}
+
+impl<'a, R: BleAdvertisementDriver + 'a, A: hil::time::Alarm + 'a> hil::time::Client
+    for BLE<'a, R, A> {
+    // this method is called once the virtual timer has been expired
+    // used to periodically send BLE advertisements without blocking the kernel
+    fn fired(&self) {
+        if self.advertise.get() == true {
+            self.radio.start_adv();
+        } else {
+            self.radio.continue_adv();
+        }
+    }
+}
+
+impl<'a, R: BleAdvertisementDriver + 'a, A: hil::time::Alarm + 'a> Client for BLE<'a, R, A> {
+    fn continue_adv(&self) {
+        self.advertise.set(false);
+        let tics = self.alarm.now().wrapping_add(2 as u32);
```

Comment:
Yes, we wanted the chip module to handle this. While developing we had debug prints between each transmission, this added enough delay to make it all work. When removing the "delay" it didn't work any longer. Setting the delay to 2 in the capsule made it work again. 

One suggestion would be to use a delay inside the chip module from a the "delay" capsule.